### PR TITLE
Upgrade to Rails 4.2

### LINKED ISSUE #1775: Upgrade to Rails 4.2
See http://guides.rubyonrails.org/4_2_release_notes.html#upgrading-to-rails-4-2

### PR DESCRIPTION
**Pull request by dnrce**
Resolves #1775

### CODE REVIEW

**Comment by dnrce** [comment]
Letting Travis do the testing work for me. :-)

Still needs some work -- closing until I revisit.

**Comment by dnrce** [comment]
I'm puzzled by the failures of `assert_select_encoded`. Anyone else is welcome to take a look.

**Comment by mattr-** [comment]
Looks like some tests are still failing. Are those legit errors that we should try and fix?

**Comment by dnrce** [comment]
These are legit issues. Some are only with the tests themselves. The URL whitelist issue also needs to be addressed.

I reopened this so I could push some rebased changes, and to get more eyes on it. Not ready to merge.

**Comment by dnrce** [comment]
Opened rails/rails-dom-testing#42 regarding the `assert_select_encoded` issues.

**Comment by mattr-** [comment]
Can we get a rebase on this? I'm rerunning the tests on this locally and would love to merge it soon. If it's not something that you can get to soon, I can pull it locally and rebase it.

**Comment by dnrce** [comment]
Updated. I had a few test changes in my local branch so this might not be 100% up to date, but it resolves the conflict.

Still not passing.

**Comment by dnrce** [comment]
Filed another bug against rails-dom-testing that is causing `assert_select` to fail on the Atom feeds: rails/rails-dom-testing#51

**Comment by dnrce** [comment]
And another one: rails/rails-dom-testing#52

**Comment by dnrce** [comment]
After fighting through/working around all the annoying behavior changes in `rails-dom-testing`, I think this is ready to go!
* The unit tests are now passing.
* The feature tests have a slight regression from 5 to 6 failures but they're all the same type so IMO it's probably part of the existing set of issues stemming from the UI changes. The one new failure is `features/recurring_todos.feature:15`.

It's only taken me nine months. But hey, I beat the release of Rails 5. :smile:

**Comment by dnrce** [comment]
Actually, doing a little bit of squashing..

**Comment by dnrce** [comment]
Squashed from 16 to 8 commits.

**Comment by mattr-** [comment]
Cool. Thanks for working on this! I'll take a look soon.

**Comment by dnrce** [comment]
@mattr-, any chance to review this yet? I'm just hoping to get this in before Rails 5 makes this officially obsolete. :smile:

File: config/environments/development.rb [hunk 1]
```diff
@@ -27,6 +27,10 @@
   # number of complex assets.
   config.assets.debug = true
 
+  # Asset digests allow you to set far-future HTTP expiration dates on all assets,
+  # yet still be able to expire them through the digest params.
+  config.assets.digest = true
```
Comment (mattr-) [inline config/environments/development.rb]: i'm not sure we want to enable digests in development mode, do we?

Comment (dnrce) [inline config/environments/development.rb]: It's to provide better parity between environments. See rails/rails#15155

Comment (mattr-) [inline config/environments/development.rb]: :+1:

**Comment by mattr-** [comment]
This looks good to me.

**Comment by mattr-** [comment]
Forgot to merge this the other day. 😞

**Comment by dnrce** [comment]
Woohoo! Thanks, @mattr-!

**Pull request merged**